[Community Pipelines]Accelerate inference of stable diffusion xl (SDXL) by IPEX on CPU (#6683)
huggingface/diffusers: examples/community/README.md

```

### Stable Diffusion XL on IPEX

This diffusion pipeline aims to accelarate the inference of Stable-Diffusion XL on Intel Xeon CPUs with BF16/FP32 precision using [IPEX](https://github.com/intel/intel-extension-for-pytorch).

To use this pipeline, you need to:
1. Install [IPEX](https://github.com/intel/intel-extension-for-pytorch)

> **Note:** For each PyTorch release, there is a corresponding release of the IPEX. Here is the mapping relationship. It is recommended to install Pytorch/IPEX2.0 to get the best performance.

Reviewer: ```suggestion
**Note:** For each PyTorch release, there is a corresponding release of IPEX. Here is the mapping relationship. It is recommended to install Pytorch/IPEX2.0 to get the best performance.
```

|PyTorch Version|IPEX Version|
|--|--|
|[v2.0.\*](https://github.com/pytorch/pytorch/tree/v2.0.1 "v2.0.1")|[v2.0.\*](https://github.com/intel/intel-extension-for-pytorch/tree/v2.0.100+cpu)|
|[v1.13.\*](https://github.com/pytorch/pytorch/tree/v1.13.0 "v1.13.0")|[v1.13.\*](https://github.com/intel/intel-extension-for-pytorch/tree/v1.13.100+cpu)|

You can simply use pip to install IPEX with the latest version.
```python
python -m pip install intel_extension_for_pytorch
```
**Note:** To install a specific version, run with the following command:
```
python -m pip install intel_extension_for_pytorch==<version_name> -f https://developer.intel.com/ipex-whl-stable-cpu
```

2. After pipeline initialization, `prepare_for_ipex()` should be called to enable IPEX accelaration. Supported inference datatypes are Float32 and BFloat16.

> **Note:** The setting of generated image height/width for `prepare_for_ipex()` should be same as the setting of pipeline inference.

Reviewer: ```suggestion
**Note:** The setting of generated image height/width for `prepare_for_ipex()` should be same as the setting of pipeline inference.

```

> **Note:** The setting of generated image height/width for `prepare_for_ipex()` should be same as the setting of pipeline inference.

Reviewer: ```suggestion
**Note:** The values of `height` and `width` used during preparation with `prepare_for_ipex()` should be the same when running inference with the prepared pipeline.
```

```python
pipe = StableDiffusionXLPipelineIpex.from_pretrained("stabilityai/sdxl-turbo", use_auth_token=True, low_cpu_mem_usage=True, use_safetensors=True)
# For Float32
pipe.prepare_for_ipex(data_type, prompt, height=512, width=512) #value of image height/width should be consistent with the pipeline inference
# For BFloat16

> pipe.prepare_for_ipex(data_type, prompt, height=512, width=512) #value of image height/width should be consistent with the pipeline inference

Reviewer: The two lines are identical. Maybe:
- Spell out the data types `torch.bfloat16`, `torch.float32` instead of using the same variable name.
- Remove the long comments in the invocation lines, we could put it once at the beginning of the snippet.

```

Then you can use the ipex pipeline in a similar way to the default stable diffusion xl pipeline.
```python
# For Float32
image = pipe(prompt, num_inference_steps=num_inference_steps, height=512, width=512, guidance_scale=guidance_scale).images[0] #value of image height/width should be consistent with 'prepare_for_ipex()'
# For BFloat16
with torch.cpu.amp.autocast(enabled=True, dtype=torch.bfloat16):

>     image = pipe(prompt, num_inference_steps=num_inference_steps, height=512, width=512, guidance_scale=guidance_scale).images[0] #value of image height/width should be consistent with 'prepare_for_ipex()'

Reviewer: Same here, the long lines with the comments can be distracting

```

> The following code compares the performance of the original stable diffusion xl pipeline with the ipex-optimized pipeline.

Reviewer: Can we get a brief qualitative comment about the kind of speedup we should expect?

Author: Hi @pcuenca, thanks so much for the detailed suggestions. I have modified all parts as you suggested. Except for some wording part, most important modifications are:
1. `data_type` change to `torch.float32`/`torch.bfloat16` accordingly
2. `use_auth_token` is not needed in this pipeline, so I simply remove all of it
3. give an estimated performance boost (1.4-2x) in the comment, which is tested using BFloat16 on the fourth generation of Intel Xeon CPU code-named SPR
4. remove the long-line comments and make it appear only once at the beginning
Pls help to review the updated code, thanks!

```python
import torch
from diffusers import StableDiffusionXLPipeline
from pipeline_stable_diffusion_xl_ipex import StableDiffusionXLPipelineIpex
import time

prompt = "sailing ship in storm by Rembrandt"
model_id = "stabilityai/sdxl-turbo"
steps = 4

# Helper function for time evaluation
def elapsed_time(pipeline, nb_pass=3, num_inference_steps=1):
    # warmup
    for _ in range(2):
        images = pipeline(prompt, num_inference_steps=num_inference_steps, height=512, width=512, guidance_scale=0.0).images
    #time evaluation
    start = time.time()
    for _ in range(nb_pass):
        pipeline(prompt, num_inference_steps=num_inference_steps, height=512, width=512, guidance_scale=0.0)
    end = time.time()
    return (end - start) / nb_pass

##############     bf16 inference performance    ###############

# 1. IPEX Pipeline initialization

> pipe = StableDiffusionXLPipelineIpex.from_pretrained(model_id, use_auth_token=True, low_cpu_mem_usage=True, use_safetensors=True)

Reviewer: ```suggestion
pipe = StableDiffusionXLPipelineIpex.from_pretrained(model_id, low_cpu_mem_usage=True, use_safetensors=True)
```

Do we need `use_auth_token`?
